lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package storage
+
+import (
+	"bytes"
+	"maps"
+	"sync"
+
+	xmaps "golang.org/x/exp/maps"
+
+	"github.com/ChainSafe/gossamer/pkg/trie"
+	"github.com/ChainSafe/gossamer/pkg/trie/codec"
+)
+
+type storageDiff struct {
+	upserts        map[string][]byte
+	deletes        map[string]bool
+	childChangeSet map[string]*storageDiff
+	l              sync.RWMutex
+}
+
+func newChangeSet() *storageDiff {
+	return &storageDiff{
+		upserts:        make(map[string][]byte),
+		deletes:        make(map[string]bool),
+		childChangeSet: make(map[string]*storageDiff),
+		l:              sync.RWMutex{},
+	}
+}
+
+func (cs *storageDiff) get(key string) ([]byte, bool) {
+	if cs == nil {
+		return nil, false
+	}
+
+	cs.l.RLock()
+	defer cs.l.RUnlock()
+
+	// Check in recent upserts if not found check if we want to delete it
+	if val, ok := cs.upserts[key]; ok {
+		return val, false
+	} else if deleted := cs.deletes[key]; deleted {
+		return nil, true
+	}
+
+	return nil, false
+}
+
+func (cs *storageDiff) upsert(key string, value []byte) {
+	if cs == nil {
+		return
+	}
+
+	cs.l.Lock()
+	defer cs.l.Unlock()
+	// If we previously deleted this trie we have to undo that deletion
+	if cs.deletes[key] {
+		delete(cs.deletes, key)
+	}
+
+	cs.upserts[key] = value
+}
+
+func (cs *storageDiff) delete(key string) {
+	if cs == nil {
+		return
+	}
+
+	cs.l.Lock()
+	defer cs.l.Unlock()
+
+	delete(cs.childChangeSet, key)
+	delete(cs.upserts, key)
+	cs.deletes[key] = true
+}
+
+func (cs *storageDiff) clearPrefix(prefix []byte, trieKeys []string, limit int) (deleted uint32, allDeleted bool) {
+	// TODO: consider actual keys in trie
+	prefix = codec.KeyLEToNibbles(prefix)
+	prefix = bytes.TrimSuffix(prefix, []byte{0})
+	newKeys := xmaps.Keys(cs.upserts)
+	allKeys := append(newKeys, trieKeys...)
+	deleted = 0
+	for _, k := range allKeys {
+		if limit > -1 && deleted >= uint32(limit) {
```

Comment:
I think you do this check outside of the for loop since we know deleted's max value?